### ISSUE TITLE
android on_resume: _process called before enter_vr_mode()

### ISSUE THREAD

**Issue by NeoSpark314** [GodotVR/godot_oculus_mobile#50]
I currently have the issue that when the app is paused and then resumed again that
gdscript `_process()` is called once before ovr is initialized again (via `OvrMobileSession::enter_vr_mode()`
This means that for one frame all calls to the gdns api functions are invalid.

Since ovr is recreated on app resume all settings need to be set again as they are initialized with the defaults. I would like to set them for example in a gdscript function that reacts on app resume:

```
func _notification(what):
	if (what == NOTIFICATION_APP_PAUSED):
		pass;
	if (what == NOTIFICATION_APP_RESUMED):
		set_all_app_settings(); # problem: ovr invalid at this point
```

A workaround would maybe be to set on NOTIFICATION_APP_RESUMED a counter; wait for >= frame in _process() and then set the settings; but then one frame would be rendered with "wrong" settings; and also all game logic would need to check if accessing gdns vrApi functions is valid.

Another Idea is maybe to call enter_vr for each exposed api function when not initialized?

**Comment by NeoSpark314** [comment]
I also just noticed that _process() is called once after NOTIFICATION_APP_PAUSED before the app is actually paused.

**Comment by m4gr3d** [comment]
@NeoSpark314 Do you have the latest from upstream for both the oculus mobile plugin and the godot engine?

The logic in [GodotRenderer#onDrawFrame](https://github.com/godotengine/godot/blob/3b532aa7ebddd47f8b041190c30b5b473dfbe116/platform/android/java/lib/src/org/godotengine/godot/GodotRenderer.java#L45) should enforce that `NOTIFICATION_APP_RESUMED` is invoked before `_process` since `GodotLib.onRendererResumed` is invoked beofre `GodotLib.step()`. If you're seeing them out of order, that would indicate a bug in the way that logic is handled in the godot engine.

What you mention about `NOTIFICATION_APP_PAUSED` is possible however. Do you see it happening all the time, or randomly?

**Comment by NeoSpark314** [comment]
@m4gr3d Thanks; that was part of the problem; I just redid the test with the master from 1 hour ago: c11ccb7e6
I noticed that I need to call
`./gradlew build` before `./gradlew generateGodotTemplates`; else it just takes the previously build libs instead of calling scons to build the changed cpp files again.

The second problem I mentioned seems to be gone now; process is not called after NOTIFICATION_APP_PAUSED; The first problem is almost gone; when process is called after resume ovr is initialized; but it is not initialized inside the _notification() yet when APP_RESUMED is notified.

The problem is visible in line `09-30 21:00:59.891` of the adb logcat below. I use the call to `get_supported_display_refresh_rates()` to check if calls to ovr are valid.

Here is my test script and the result:

```
extends Spatial

var ovrDisplayRefreshRate = null;

func _notification(what):
	if (what == NOTIFICATION_APP_PAUSED):
		print("NOTIFICATION_APP_PAUSED");
		print(ovrDisplayRefreshRate.get_supported_display_refresh_rates());
	if (what == NOTIFICATION_APP_RESUMED):
		print("NOTIFICATION_APP_RESUMED");
		print(ovrDisplayRefreshRate.get_supported_display_refresh_rates());
		
		
func _process(dt):
	print("_process()");
	print(ovrDisplayRefreshRate.get_supported_display_refresh_rates())
	

func _ready():
	var arvr_interface = ARVRServer.find_interface("OVRMobile")
	if arvr_interface and arvr_interface.initialize():
		get_viewport().arvr = true
		Engine.target_fps = 72

		var OvrDisplayRefreshRate = load("res://addons/godot_ovrmobile/OvrDisplayRefreshRate.gdns");
		ovrDisplayRefreshRate = OvrDisplayRefreshRate.new();
```

And the output of logcat:
```
$ adb logcat -s godot:* GodotOVRMobile:*
--------- beginning of crash
--------- beginning of system
--------- beginning of main
09-30 21:00:47.033 30593 30609 I godot   : Godot Engine v3.2.alpha.custom_build.c11ccb7e6 - https://godotengine.org
09-30 21:00:47.041 30593 30609 I godot   : OpenGL ES 2.0 Renderer: Adreno (TM) 540
09-30 21:00:47.083 30593 30609 I godot   :  
09-30 21:00:47.207 30593 30609 V GodotOVRMobile: Creating OVR Mobile session.
09-30 21:00:47.308 30593 30609 V GodotOVRMobile: OvrMobileSession::initialize() called
09-30 21:00:47.505 30593 30609 V GodotOVRMobile:  render target size multiplier: 1.000000
09-30 21:00:47.505 30593 30609 V GodotOVRMobile:  vrapi render target size: w 1216 / h 1344
09-30 21:00:47.519 30593 30609 V GodotOVRMobile:     Create Framebuffer(...): mTextureSwapChainLength=3
09-30 21:00:47.533 30593 30609 V GodotOVRMobile:     Create Framebuffer(...): mTextureSwapChainLength=3
09-30 21:00:47.652 30593 30609 V GodotOVRMobile: Entered Oculus Mobile VR mode.
09-30 21:00:47.652 30593 30609 V GodotOVRMobile: 		vrapi_SetPerfThread( RENDERER, 30609 )
09-30 21:00:48.513 30593 30609 I godot   : _process()
09-30 21:00:48.513 30593 30609 I godot   : [60, 72]
09-30 21:00:48.650 30593 30609 I godot   : _process()
09-30 21:00:48.650 30593 30609 I godot   : [60, 72]
09-30 21:00:48.659 30593 30609 I godot   : _process()
09-30 21:00:48.659 30593 30609 I godot   : [60, 72]
.
.
.
09-30 21:00:50.244 30593 30609 I godot   : _process()
09-30 21:00:50.244 30593 30609 I godot   : [60, 72]
09-30 21:00:50.273 30593 30609 I godot   : NOTIFICATION_APP_PAUSED
09-30 21:00:50.273 30593 30609 I godot   : [60, 72]
09-30 21:00:50.487 30593 30609 V GodotOVRMobile: Left Oculus Mobile VR mode.
.
.
.
09-30 21:00:59.891 30593 30609 I godot   : NOTIFICATION_APP_RESUMED
09-30 21:00:59.891 30593 30609 I godot   : Null
09-30 21:01:00.033 30593 30609 V GodotOVRMobile: Entered Oculus Mobile VR mode.
09-30 21:01:00.033 30593 30609 V GodotOVRMobile: 		vrapi_SetPerfThread( RENDERER, 30609 )
09-30 21:01:00.034 30593 30609 I godot   : _process()
09-30 21:01:00.034 30593 30609 I godot   : [60, 72]
09-30 21:01:00.044 30593 30609 I godot   : _process()
09-30 21:01:00.044 30593 30609 I godot   : [60, 72]
09-30 21:01:00.055 30593 30609 I godot   : _process()
...
```

But with this now it is possible to reset all settings at the beginning of _process() after an APP_RESUMED notification without affecting the game logic.

**Comment by m4gr3d** [comment]
@NeoSpark314 I removed the logic that used to call `scons` automatically in [PR #32269](https://github.com/godotengine/godot/pull/32269) based on feedback in that pull request.
So for the proper behavior, you now have to manually call `scons` to generate the desired native shared library before invoking `gradlew generateGodotTemplates`.
I've added more information in that PR description, and have [PR #2796](https://github.com/godotengine/godot-docs/pull/2796) in flight to update the official documentation.

Regarding the second observation, it's correct but it's working as intended as the plugin uses the `NOTIFICATION_APP_RESUMED` event to reinitialize ovr, ensuring that it's valid whenever invoked within the `_process` callback.
A workaround to the logic you have would be to trigger a flag whenever you receive the `NOTIFICATION_APP_RESUMED` event and check that flag in `_process` to figure out when is a good time to reset the settings.
